orchestrator/utils/search_query.py

```diff
@@ -0,0 +1,348 @@
+    @staticmethod
+    def visit_group(node: Node, acc: list[str]) -> None:
+        acc.append("(")
+        for expr in node[1]:
+            TSQueryVisitor.visit_and_expression(expr, acc)
+            acc.append(" | ")
+        acc.pop()
+        acc.append(")")
+
+    @staticmethod
+    def visit_kv_term(node: Node, acc: list[str]) -> None:
+        key_node, value_node = node[1]
+
+        # Re-use visit_term
+        TSQueryVisitor.visit_term(key_node, acc)
+        acc.append(" <-> ")
+
+        if value_node[0] == "ValueGroup":
+            acc.append("(")
+            for v in value_node[1]:
+                TSQueryVisitor.visit_search_word(v, acc)
+                acc.append(" | ")
+            acc.pop()
+            acc.append(")")
+        else:
+            TSQueryVisitor.visit_term(value_node, acc)
+
+    @staticmethod
+    def visit_search_word(node: Node, acc: list[str]) -> None:
+        # Postgres is finicky with single quotes in the query. In some situations it throws an error.
+        # To avoid the special handling of ' by PG, we replace it with double quotes.
+        text = node[1].replace("'", '"')
+        if node[0] == "Word":
+            acc.append(text)
+        elif node[0] == "PrefixWord":
+            acc.append(f"{text}:*")
+        else:
+            raise Exception(f"Invalid SearchWord type {node[0]}")
+
+    @staticmethod
+    def visit_phrase(node: Node, acc: list[str]) -> None:
+        for search_word in node[1]:
+            TSQueryVisitor.visit_search_word(search_word, acc)
+            acc.append(" <-> ")
+
+        if len(node[1]):
+            acc.pop()  # Remove trailing space
+
+    @staticmethod
+    def visit_term(node: Node, acc: list[str]) -> None:
+        node_type = node[0]
+        if node_type == "Phrase":
+            TSQueryVisitor.visit_phrase(node, acc)
+        elif node_type in ["Word", "PrefixWord"]:
+            TSQueryVisitor.visit_search_word(node, acc)
+        elif node_type == "KVTerm":
+            TSQueryVisitor.visit_kv_term(node, acc)
+        elif node_type == "Group":
+            TSQueryVisitor.visit_group(node, acc)
+        elif node_type == "Negation":
+            acc.append("!")
+            should_group = node[1][0] in ["Phrase", "KVTerm"]
+            if should_group:
+                acc.append("(")
+            TSQueryVisitor.visit_term(node[1], acc)
+            if should_group:
+                acc.append(")")
+        else:
+            raise Exception(f"Unexpected term node type: {node_type}")
+
+    @staticmethod
+    def visit_and_expression(node: Node, acc: list[str]) -> None:
+        for term in node[1]:
+            TSQueryVisitor.visit_term(term, acc)
+            acc.append(" & ")
+        if len(node[1]):
+            acc.pop()
+
+    @staticmethod
+    def visit_query(node: Node, acc: list[str]) -> None:
+        for expression in node[1]:
+            TSQueryVisitor.visit_and_expression(expression, acc)
+            acc.append(" | ")
```

Comment:
I understand that `AND` is the default, but is it currently possible to do a query `foo OR bar`?